spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkTable.java

```diff
@@ -247,6 +248,11 @@ public ScanBuilder newScanBuilder(CaseInsensitiveStringMap options) {
 
   @Override
   public WriteBuilder newWriteBuilder(LogicalWriteInfo info) {
+    boolean branchOptionPresent = info.options().containsKey(SparkWriteOptions.BRANCH);
+    if (!branchOptionPresent) {
+      Preconditions.checkArgument(
+          snapshotId == null, "Cannot write to table at a specific snapshot: %s", snapshotId);
+    }
```

Comment:
I'll need to double check the code in case there's any other implications to this but in case we do go this route could we combine the boolean and precondition check

```
boolean branchWrite = info.options().containsKey(SparkWriteOptions.BRANCH);
Preconditions.checkArgument(branchWrite || snapshotId == null, "Cannot write to table at a specific snapshot: %s", snapshotId)
```